In [2]:
import pandas as pd
import numpy as np
from scipy import stats

## Q1.1
Data cleaning

In [3]:
raw = pd.read_csv("adult.data", names = ["age", "workclass", "fnlwgt", "education", "education-num", "marital-status", 
                                         "occupation", "relationship",
                                         "race", "sex", "capital-gain", "capital-loss", "hours-per-week", "native-country"
                                        , "income"])
raw = raw.replace(" ?", np.nan)
raw = raw.dropna()
raw

age          workclass  fnlwgt    education  education-num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[30162 rows x 15 columns]

In [5]:
raw = raw[raw["native-country"] != " Holand-Netherlands"]
raw = raw[raw["occupation"] != " Armed-Forces"]

In [6]:
dummy_fields = ["workclass", "education", "marital-status", 
                                         "occupation", "relationship",
                                         "race", "sex", "native-country", "income"]

df = raw

for each in dummy_fields:
    dummies = pd.get_dummies( df.loc[:, each], prefix=each ) 
    df = pd.concat( [df, dummies], axis = 1 )
    
fields_to_drop = dummy_fields
fields_to_drop.append("income_ <=50K")


df = df.drop( fields_to_drop, axis = 1 )
df

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
0       39   77516             13          2174             0              40   
1       50   83311             13             0             0              13   
2       38  215646              9             0             0              40   
3       53  234721              7             0             0              40   
4       28  338409             13             0             0              40   
...    ...     ...            ...           ...           ...             ...   
32556   27  257302             12             0             0              38   
32557   40  154374              9             0             0              40   
32558   58  151910              9             0             0              40   
32559   22  201490              9             0             0              20   
32560   52  287927              9         15024             0              40   

       workclass_ Federal-gov  workclass_ Local-gov  workclass_ Private  \
0                           0                     0                   0   
1                           0                     0                   0   
2                           0                     0                   1   
3                           0                     0                   1   
4                           0                     0                   1   
...                       ...                   ...                 ...   
32556                       0                     0                   1   
32557                       0                     0                   1   
32558                       0                     0                   1   
32559                       0                     0                   1   
32560                       0                     0                   0   

       workclass_ Self-emp-inc  ...  native-country_ Puerto-Rico  \
0                            0  ...                            0   
1                            0  ...                            0   
2                            0  ...                            0   
3                            0  ...                            0   
4                            0  ...                            0   
...                        ...  ...                          ...   
32556                        0  ...                            0   
32557                        0  ...                            0   
32558                        0  ...                            0   
32559                        0  ...                            0   
32560                        1  ...                            0   

       native-country_ Scotland  native-country_ South  \
0                             0                      0   
1                             0                      0   
2                             0                      0   
3                             0                      0   
4                             0                      0   
...                         ...                    ...   
32556                         0                      0   
32557                         0                      0   
32558                         0                      0   
32559                         0                      0   
32560                         0                      0   

       native-country_ Taiwan  native-country_ Thailand  \
0                           0                         0   
1                           0                         0   
2                           0                         0   
3                           0                         0   
4                           0                         0   
...                       ...                       ...   
32556                       0                         0   
32557                       0                         0   
32558                       0                         0   
32559                       0

In [7]:
df.columns[df.sum(axis=0) <= 10]

Index([], dtype='object')

In [8]:
#df = df.drop(["occupation_ Armed-Forces", "native-country_ Holand-Netherlands"], axis = 1 )
df = df.reset_index()
df

index  age  fnlwgt  education-num  capital-gain  capital-loss  \
0          0   39   77516             13          2174             0   
1          1   50   83311             13             0             0   
2          2   38  215646              9             0             0   
3          3   53  234721              7             0             0   
4          4   28  338409             13             0             0   
...      ...  ...     ...            ...           ...           ...   
30147  32556   27  257302             12             0             0   
30148  32557   40  154374              9             0             0   
30149  32558   58  151910              9             0             0   
30150  32559   22  201490              9             0             0   
30151  32560   52  287927              9         15024             0   

       hours-per-week  workclass_ Federal-gov  workclass_ Local-gov  \
0                  40                       0                     0   
1                  13                       0                     0   
2                  40                       0                     0   
3                  40                       0                     0   
4                  40                       0                     0   
...               ...                     ...                   ...   
30147              38                       0                     0   
30148              40                       0                     0   
30149              40                       0                     0   
30150              20                       0                     0   
30151              40                       0                     0   

       workclass_ Private  ...  native-country_ Puerto-Rico  \
0                       0  ...                            0   
1                       0  ...                            0   
2                       1  ...                            0   
3                       1  ...                            0   
4                       1  ...                            0   
...                   ...  ...                          ...   
30147                   1  ...                            0   
30148                   1  ...                            0   
30149                   1  ...                            0   
30150                   1  ...                            0   
30151                   0  ...                            0   

       native-country_ Scotland  native-country_ South  \
0                             0                      0   
1                             0                      0   
2                             0                      0   
3                             0                      0   
4                             0                      0   
...                         ...                    ...   
30147                         0                      0   
30148                         0                      0   
30149                         0                      0   
30150                         0                      0   
30151                         0                      0   

       native-country_ Taiwan  native-country_ Thailand  \
0                           0                         0   
1                           0                         0   
2                           0                         0   
3                           0                         0   
4                           0                         0   
...                       ...                       ...   
30147                       0                         0   
30148                       0                         0   
30149                       0                         0   
30150                       0                         0   
30151                       0                         0   

       native-country_ Trinadad&Tobago  native-country_ United-States  \
0                                    0               

In [9]:
test = pd.read_csv("adult.test", names = ["age", "workclass", "fnlwgt", "education", "education-num", "marital-status", 
                                         "occupation", "relationship",
                                         "race", "sex", "capital-gain", "capital-loss", "hours-per-week", "native-country"
                                        , "income"])
test = test.replace(" ?", np.nan)
test = test.dropna()
test

age      workclass    fnlwgt      education  education-num  \
1      25        Private  226802.0           11th            7.0   
2      38        Private   89814.0        HS-grad            9.0   
3      28      Local-gov  336951.0     Assoc-acdm           12.0   
4      44        Private  160323.0   Some-college           10.0   
6      34        Private  198693.0           10th            6.0   
...    ..            ...       ...            ...            ...   
16276  33        Private  245211.0      Bachelors           13.0   
16277  39        Private  215419.0      Bachelors           13.0   
16279  38        Private  374983.0      Bachelors           13.0   
16280  44        Private   83891.0      Bachelors           13.0   
16281  35   Self-emp-inc  182148.0      Bachelors           13.0   

            marital-status          occupation    relationship  \
1            Never-married   Machine-op-inspct       Own-child   
2       Married-civ-spouse     Farming-fishing         Husband   
3       Married-civ-spouse     Protective-serv         Husband   
4       Married-civ-spouse   Machine-op-inspct         Husband   
6            Never-married       Other-service   Not-in-family   
...                    ...                 ...             ...   
16276        Never-married      Prof-specialty       Own-child   
16277             Divorced      Prof-specialty   Not-in-family   
16279   Married-civ-spouse      Prof-specialty         Husband   
16280             Divorced        Adm-clerical       Own-child   
16281   Married-civ-spouse     Exec-managerial         Husband   

                      race      sex  capital-gain  capital-loss  \
1                    Black     Male           0.0           0.0   
2                    White     Male           0.0           0.0   
3                    White     Male           0.0           0.0   
4                    Black     Male        7688.0           0.0   
6                    White     Male           0.0           0.0   
...                    ...      ...           ...           ...   
16276                White     Male           0.0           0.0   
16277                White   Female           0.0           0.0   
16279                White     Male           0.0           0.0   
16280   Asian-Pac-Islander     Male        5455.0           0.0   
16281                White     Male           0.0           0.0   

       hours-per-week  native-country   income  
1                40.0   United-States   <=50K.  
2                50.0   United-States   <=50K.  
3                40.0   United-States    >50K.  
4                40.0   United-States    >50K.  
6                30.0   United-States   <=50K.  
...               ...             ...      ...  
16276            40.0   United-States   <=50K.  
16277            36.0   United-States   <=50K.  
16279            50.0   United-States   <=50K.  
16280            40.0   United-States   <=50K.  
16281            60.0   United-States    >50K.  

[15060 rows x 15 columns]

In [10]:
test = test[test["native-country"] != " Holand-Netherlands"]
test = test[test["occupation"] != " Armed-Forces"]

In [11]:
dummy_fields = ["workclass", "education", "marital-status", 
                                         "occupation", "relationship",
                                         "race", "sex", "native-country", "income"]

test.df = test

for each in dummy_fields:
    dummies = pd.get_dummies( test.df.loc[:, each], prefix=each ) 
    test.df = pd.concat( [test.df, dummies], axis = 1 )
    
fields_to_drop = dummy_fields
fields_to_drop.append("income_ <=50K.")

test.df = test.df.drop(fields_to_drop, axis = 1)
test.df

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  """


age    fnlwgt  education-num  capital-gain  capital-loss  \
1      25  226802.0            7.0           0.0           0.0   
2      38   89814.0            9.0           0.0           0.0   
3      28  336951.0           12.0           0.0           0.0   
4      44  160323.0           10.0        7688.0           0.0   
6      34  198693.0            6.0           0.0           0.0   
...    ..       ...            ...           ...           ...   
16276  33  245211.0           13.0           0.0           0.0   
16277  39  215419.0           13.0           0.0           0.0   
16279  38  374983.0           13.0           0.0           0.0   
16280  44   83891.0           13.0        5455.0           0.0   
16281  35  182148.0           13.0           0.0           0.0   

       hours-per-week  workclass_ Federal-gov  workclass_ Local-gov  \
1                40.0                       0                     0   
2                50.0                       0                     0   
3                40.0                       0                     1   
4                40.0                       0                     0   
6                30.0                       0                     0   
...               ...                     ...                   ...   
16276            40.0                       0                     0   
16277            36.0                       0                     0   
16279            50.0                       0                     0   
16280            40.0                       0                     0   
16281            60.0                       0                     0   

       workclass_ Private  workclass_ Self-emp-inc  ...  \
1                       1                        0  ...   
2                       1                        0  ...   
3                       0                        0  ...   
4                       1                        0  ...   
6                       1                        0  ...   
...                   ...                      ...  ...   
16276                   1                        0  ...   
16277                   1                        0  ...   
16279                   1                        0  ...   
16280                   1                        0  ...   
16281                   0                        1  ...   

       native-country_ Puerto-Rico  native-country_ Scotland  \
1                                0                         0   
2                                0                         0   
3                                0                         0   
4                                0                         0   
6                                0                         0   
...                            ...                       ...   
16276                            0                         0   
16277                            0                         0   
16279                            0                         0   
16280                            0                         0   
16281                            0                         0   

       native-country_ South  native-country_ Taiwan  \
1                          0                       0   
2                          0                       0   
3                          0                       0   
4                          0                       0   
6                          0                       0   
...                      ...                     ...   
16276                      0                       0   
16277                      0                       0   
16279                      0                       0   
16280                      0                       0   
16281                      0                       0   

       native-country_ Thailand  native-country_ Trinadad&Tobago  \
1                             0                                0   
2                             0                                0   
3    

In [12]:
#test.df = test.df.drop(["occupation_ Armed-Forces"], axis = 1 )
test.df = test.df.reset_index()
test.df

index age    fnlwgt  education-num  capital-gain  capital-loss  \
0          1  25  226802.0            7.0           0.0           0.0   
1          2  38   89814.0            9.0           0.0           0.0   
2          3  28  336951.0           12.0           0.0           0.0   
3          4  44  160323.0           10.0        7688.0           0.0   
4          6  34  198693.0            6.0           0.0           0.0   
...      ...  ..       ...            ...           ...           ...   
15050  16276  33  245211.0           13.0           0.0           0.0   
15051  16277  39  215419.0           13.0           0.0           0.0   
15052  16279  38  374983.0           13.0           0.0           0.0   
15053  16280  44   83891.0           13.0        5455.0           0.0   
15054  16281  35  182148.0           13.0           0.0           0.0   

       hours-per-week  workclass_ Federal-gov  workclass_ Local-gov  \
0                40.0                       0                     0   
1                50.0                       0                     0   
2                40.0                       0                     1   
3                40.0                       0                     0   
4                30.0                       0                     0   
...               ...                     ...                   ...   
15050            40.0                       0                     0   
15051            36.0                       0                     0   
15052            50.0                       0                     0   
15053            40.0                       0                     0   
15054            60.0                       0                     0   

       workclass_ Private  ...  native-country_ Puerto-Rico  \
0                       1  ...                            0   
1                       1  ...                            0   
2                       0  ...                            0   
3                       1  ...                            0   
4                       1  ...                            0   
...                   ...  ...                          ...   
15050                   1  ...                            0   
15051                   1  ...                            0   
15052                   1  ...                            0   
15053                   1  ...                            0   
15054                   0  ...                            0   

       native-country_ Scotland  native-country_ South  \
0                             0                      0   
1                             0                      0   
2                             0                      0   
3                             0                      0   
4                             0                      0   
...                         ...                    ...   
15050                         0                      0   
15051                         0                      0   
15052                         0                      0   
15053                         0                      0   
15054                         0                      0   

       native-country_ Taiwan  native-country_ Thailand  \
0                           0                         0   
1                           0                         0   
2                           0                         0   
3                           0                         0   
4                           0                         0   
...                       ...                       ...   
15050                       0                         0   
15051                       0                         0   
15052                       0                         0   
15053                       0                         0   
15054                       0                         0   

       native-country_ Trinadad&Tobago  native-country_ United-States  \
0                                    0   

In [13]:
test.df['age'] = test.df['age'].astype(int)

In [14]:
train_cols = pd.DataFrame(df.columns)
test_cols = pd.DataFrame(test.df.columns)
pd.set_option('display.max_rows', 150)
pd.merge(train_cols, test_cols, how='outer', indicator=True)

0      _merge
0                                         index        both
1                                           age        both
2                                        fnlwgt        both
3                                 education-num        both
4                                  capital-gain        both
5                                  capital-loss        both
6                                hours-per-week        both
7                        workclass_ Federal-gov        both
8                          workclass_ Local-gov        both
9                            workclass_ Private        both
10                      workclass_ Self-emp-inc        both
11                  workclass_ Self-emp-not-inc        both
12                         workclass_ State-gov        both
13                       workclass_ Without-pay        both
14                              education_ 10th        both
15                              education_ 11th        both
16                              education_ 12th        both
17                           education_ 1st-4th        both
18                           education_ 5th-6th        both
19                           education_ 7th-8th        both
20                               education_ 9th        both
21                        education_ Assoc-acdm        both
22                         education_ Assoc-voc        both
23                         education_ Bachelors        both
24                         education_ Doctorate        both
25                           education_ HS-grad        both
26                           education_ Masters        both
27                         education_ Preschool        both
28                       education_ Prof-school        both
29                      education_ Some-college        both
30                     marital-status_ Divorced        both
31            marital-status_ Married-AF-spouse        both
32           marital-status_ Married-civ-spouse        both
33        marital-status_ Married-spouse-absent        both
34                marital-status_ Never-married        both
35                    marital-status_ Separated        both
36                      marital-status_ Widowed        both
37                     occupation_ Adm-clerical        both
38                     occupation_ Craft-repair        both
39                  occupation_ Exec-managerial        both
40                  occupation_ Farming-fishing        both
41                occupation_ Handlers-cleaners        both
42                occupation_ Machine-op-inspct        both
43                    occupation_ Other-service        both
44                  occupation_ Priv-house-serv        both
45                   occupation_ Prof-specialty        both
46                  occupation_ Protective-serv        both
47                            occupation_ Sales        both
48                     occupation_ Tech-support        both
49                 occupation_ Transport-moving        both
50                        relationship_ Husband        both
51                  relationship_ Not-in-family        both
52                 relationship_ Other-relative        both
53                      relationship_ Own-child        both
54                      relationship_ Unmarried        both
55                           relationship_ Wife        both
56                     race_ Amer-Indian-Eskimo        both
57                     race_ Asian-Pac-Islander        both
58                                  race_ Black        both
59                                  race_ Other        both
60                                  race_ White        both
61                                  sex_ Female        both
62                                    sex_ Male        both
63                     native-country_ Cambodia        both
64                       native-country_ Canada        both
65                        native-country_ China        both
66                     nat

In [326]:
x_train = np.array(df.drop(["index", "income_ >50K"], axis = 1))
x_test = np.array(test.df.drop(["index", "income_ >50K."], axis = 1))
y_train = np.array(df["income_ >50K"])
y_test = np.array(test.df["income_ >50K."])
y_test

array([0, 0, 1, ..., 0, 0, 1], dtype=uint8)

#### Summary

In [16]:
df.describe()

index           age        fnlwgt  education-num  capital-gain  \
count  30152.000000  30152.000000  3.015200e+04   30152.000000  30152.000000   
mean   16287.011508     38.440568  1.897916e+05      10.121319   1092.370025   
std     9383.452512     13.135362  1.056567e+05       2.550204   7407.547899   
min        0.000000     17.000000  1.376900e+04       1.000000      0.000000   
25%     8177.750000     28.000000  1.176248e+05       9.000000      0.000000   
50%    16284.500000     37.000000  1.784250e+05      10.000000      0.000000   
75%    24402.250000     47.000000  2.376255e+05      13.000000      0.000000   
max    32560.000000     90.000000  1.484705e+06      16.000000  99999.000000   

       capital-loss  hours-per-week  workclass_ Federal-gov  \
count  30152.000000    30152.000000            30152.000000   
mean      88.266085       40.931348                0.030976   
std      404.046306       11.979776                0.173257   
min        0.000000        1.000000                0.000000   
25%        0.000000       40.000000                0.000000   
50%        0.000000       40.000000                0.000000   
75%        0.000000       45.000000                0.000000   
max     4356.000000       99.000000                1.000000   

       workclass_ Local-gov  workclass_ Private  ...  \
count          30152.000000        30152.000000  ...   
mean               0.068553            0.739089  ...   
std                0.252696            0.439139  ...   
min                0.000000            0.000000  ...   
25%                0.000000            0.000000  ...   
50%                0.000000            1.000000  ...   
75%                0.000000            1.000000  ...   
max                1.000000            1.000000  ...   

       native-country_ Puerto-Rico  native-country_ Scotland  \
count                 30152.000000              30152.000000   
mean                      0.003615                  0.000365   
std                       0.060017                  0.019097   
min                       0.000000                  0.000000   
25%                       0.000000                  0.000000   
50%                       0.000000                  0.000000   
75%                       0.000000                  0.000000   
max                       1.000000                  1.000000   

       native-country_ South  native-country_ Taiwan  \
count           30152.000000            30152.000000   
mean                0.002355                0.001393   
std                 0.048469                0.037297   
min                 0.000000                0.000000   
25%                 0.000000                0.000000   
50%                 0.000000                0.000000   
75%                 0.000000                0.000000   
max                 1.000000                1.000000   

       native-country_ Thailand  native-country_ Trinadad&Tobago  \
count              30152.000000                     30152.000000   
mean                   0.000564                         0.000597   
std                    0.023738                         0.024426   
min                    0.000000                         0.000000   
25%                    0.000000                         0.000000   
50%                    0.000000                         0.000000   
75%                    0.000000                         0.000000   
max                    1.000000                         1.000000   

       native-country_ United-States  native-country_ Vietnam  \
count                   30152.000000             30152.000000   
mean                        0.911880                 0.002123   
std                         0.283474                 0.046023   
min                         0.000000                 0.000000   
25%                         1.000000                 0.000000   
50%                         1.000000                 0.000000   
75%                         1.000000                 0.000000   
max  

In [17]:
test.df.describe()

index           age        fnlwgt  education-num  capital-gain  \
count  15055.000000  15055.000000  1.505500e+04   15055.000000  15055.000000   
mean    8141.758087     38.769711  1.896046e+05      10.112255   1120.188907   
std     4696.092355     13.381046  1.056274e+05       2.558629   7704.274825   
min        1.000000     17.000000  1.349200e+04       1.000000      0.000000   
25%     4062.500000     28.000000  1.166450e+05       9.000000      0.000000   
50%     8146.000000     37.000000  1.779510e+05      10.000000      0.000000   
75%    12196.500000     48.000000  2.385810e+05      13.000000      0.000000   
max    16281.000000     90.000000  1.490400e+06      16.000000  99999.000000   

       capital-loss  hours-per-week  workclass_ Federal-gov  \
count  15055.000000    15055.000000            15055.000000   
mean      89.071471       40.950714                0.030422   
std      406.347469       12.064465                0.171751   
min        0.000000        1.000000                0.000000   
25%        0.000000       40.000000                0.000000   
50%        0.000000       40.000000                0.000000   
75%        0.000000       45.000000                0.000000   
max     3770.000000       99.000000                1.000000   

       workclass_ Local-gov  workclass_ Private  ...  \
count          15055.000000        15055.000000  ...   
mean               0.068615            0.732049  ...   
std                0.252807            0.442907  ...   
min                0.000000            0.000000  ...   
25%                0.000000            0.000000  ...   
50%                0.000000            1.000000  ...   
75%                0.000000            1.000000  ...   
max                1.000000            1.000000  ...   

       native-country_ Puerto-Rico  native-country_ Scotland  \
count                 15055.000000              15055.000000   
mean                      0.004384                  0.000598   
std                       0.066068                  0.024444   
min                       0.000000                  0.000000   
25%                       0.000000                  0.000000   
50%                       0.000000                  0.000000   
75%                       0.000000                  0.000000   
max                       1.000000                  1.000000   

       native-country_ South  native-country_ Taiwan  \
count           15055.000000            15055.000000   
mean                0.001993                0.000864   
std                 0.044597                0.029374   
min                 0.000000                0.000000   
25%                 0.000000                0.000000   
50%                 0.000000                0.000000   
75%                 0.000000                0.000000   
max                 1.000000                1.000000   

       native-country_ Thailand  native-country_ Trinadad&Tobago  \
count              15055.000000                     15055.000000   
mean                   0.000797                         0.000531   
std                    0.028222                         0.023046   
min                    0.000000                         0.000000   
25%                    0.000000                         0.000000   
50%                    0.000000                         0.000000   
75%                    0.000000                         0.000000   
max                    1.000000                         1.000000   

       native-country_ United-States  native-country_ Vietnam  \
count                   15055.000000             15055.000000   
mean                        0.915510                 0.001262   
std                         0.278131                 0.035504   
min                         0.000000                 0.000000   
25%                         1.000000                 0.000000   
50%                         1.000000                 0.000000   
75%                         1.000000                 0.000000   
max  

## Q1.2
Derive the gradient and hessian matrix for the new E(w).

$$\displaystyle E(w) = \frac{1} {2} w ^T \Lambda w - \sum^n_{n=1}[
t_nlny_n +(1-t_n)ln(1-y_n)]$$  
$$ y_n = \sigma(w^TX_n)$$  
$$where \space\sigma(a) = \frac{1} {1+exp(-a)} $$  
$$\frac{\partial {\sigma(a)}}{\partial a} = (-1)(1+exp(-a))^{-2} (-exp(-a))$$  
$$ = \frac{exp(-a)}{[1+exp(-a)]^2} = \sigma(a)\frac{exp(-a)}{1+exp(-a)} $$  
$$ =\sigma(a)(1-\sigma(a)) $$  

$Gradient:$  
$$\displaystyle \nabla E(w) = \frac{\partial {E(w)}}{\partial w} = \Lambda w +\sum^N_{n=1}(y_n-t_n)X_n $$  
$Hessian:$
$$\displaystyle \nabla \nabla E(w) = \Lambda+\sum^N_{n=1}y_n(1-y_n)X_nX_n^T=\Lambda+X^TRX $$  
$$ where \space R_{n,n} = y_n(1-y_n) $$

## Q1.3
Create your mylogistic_l2 class. Show the learned  𝑤  as well as test accuracy for the cases below. If  𝑤  is too long for you, show selected  𝑤  for continuous-valued, binary-valued, and the constant term.  
Case 1: lambda = 1 for all coefficients  
Case 2: lambda = 1 for all but the intercept, no regularization for intercept term.  
Case 3: lambda = 1 for numerical-valued features, lambda = 0.5 for binary-valued features, no regularization for intercept term.  

In [295]:
class mylogistic_l2(object):    
    
    def __init__(self, reg_vec = 1, max_iter = 1000, tol = 10 ** -6, add_intercept = True):
        
        self.numIterations = max_iter
        self.penalty = "L2"
        self.C = reg_vec
        self.tol = 0.1 * tol

    def train2(self, X_train, y_train):

        self.C = np.diag(self.C)
        X_train = np.c_[X_train, np.ones([np.shape(X_train)[0], 1])]
        self.weights = np.dot(np.dot(np.linalg.inv(np.dot(X_train.T, X_train) + np.mean(self.C) * np.eye(X_train.shape[1])),X_train.T),y_train)
        
        
        for i in range(self.numIterations):
            
            z = np.dot(X_train, self.weights)
            y_result = logistic_func(z)
            errors = y_result - y_train
            
            delta_w = np.dot(X_train.T, errors) + np.dot(self.C, self.weights)
            
            rnn = y_result * (1 - y_result)
            hessian = self.C + np.dot((X_train.T * rnn), X_train)
                
            self.iterationsPerformed = i
            
            delta = np.dot(np.linalg.inv(hessian), delta_w)

            if np.all(abs(delta) >= self.tol): 
                #weight update
                self.weights -= delta
            else:
                break
            
        return self
                    
    def predict(self, X_test, pi = 0.5):
       
        z = np.dot(X_test, self.weights[0:-1]) + self.weights[-1]
        probs = np.array([logistic_func(i) for i in z])
        predictions = np.where(probs >= pi, 1, 0)
       
        return predictions, probs

def logistic_func(z):   
    return 1 / (1 + np.exp(-z)) 
    

### Data standardlize

In [327]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(x_train)
x_test = scaler.transform(x_test).astype(float)
x_train = scaler.transform(x_train).astype(float)

### Case 1.

In [273]:
1e-5 *10000

0.1

In [296]:
lambda_vec = np.ones(103)
logist = mylogistic_l2(reg_vec = lambda_vec, max_iter = 100, tol = 1e-5, add_intercept = True)
logist.train2(x_train, y_train)
y_pred = logist.predict(x_test)

In [297]:
#print(logist.iterationsPerformed)
#print(logist.weights)
print("Accuracy:")
print(np.sum(y_pred[0] == y_test) / y_test.shape[0])
print("Weights:")
print(logist.weights)

Accuracy:
0.8478246429757555
Weights:
[ 3.34314569e-01  7.92912164e-02  3.76750750e-01  2.35822157e+00
  2.58537699e-01  3.52763081e-01  1.04482571e-01 -2.36736793e-02
  4.32716759e-02  5.10343051e-02 -1.07957599e-01 -4.35552516e-02
 -9.46148420e-02 -7.80664766e-02 -9.71262524e-02 -3.61490061e-02
 -2.27728707e-02 -4.14868249e-02 -1.00457356e-01 -6.93087563e-02
 -1.76577701e-02  9.80418529e-03  1.42492514e-01  1.07865663e-01
 -7.05584375e-02  1.33970367e-01 -6.10791396e-01  1.36739595e-01
  1.54005191e-02 -2.15186355e-01  5.50231362e-02  7.11030667e-01
 -6.68444852e-02 -5.15881241e-01 -1.23190953e-01 -7.15374114e-02
 -1.75841308e-02  3.32577268e-03  2.54417190e-01 -1.84552467e-01
 -1.54439997e-01 -7.79439802e-02 -2.69830202e-01 -2.21484217e-01
  1.57541646e-01  7.84570298e-02  7.77710033e-02  1.04433860e-01
 -3.20094605e-02 -3.76449469e-02  1.38497520e-01 -8.53260241e-02
 -2.89879193e-01  6.16028594e-02  2.67734326e-01 -5.06495740e-02
  5.08054485e-02 -2.67214572e-02 -3.48814576e-02  2.

### Case2.

In [298]:
lambda_vec = np.ones(103)
lambda_vec[-1] = 0
logist = mylogistic_l2(reg_vec = lambda_vec, max_iter = 100, tol = 1e-5, add_intercept = True)
logist.train2(x_train, y_train)
y_pred = logist.predict(x_test)

In [299]:
print("Accuracy:")
print(np.sum(y_pred[0] == y_test) / y_test.shape[0])
print("Weights:")
print(logist.weights)

Accuracy:
0.8477582198605115
Weights:
[ 3.34628910e-01  7.92259113e-02  3.34184297e-01  2.37333628e+00
  2.58818567e-01  3.53089015e-01  1.05047837e-01 -2.30825558e-02
  4.44404839e-02  5.14448961e-02 -1.07510137e-01 -4.30703651e-02
 -1.43714738e-01 -9.41598380e-02 -1.12516097e-01 -4.34801058e-02
 -3.48180398e-02 -5.64601256e-02 -1.18448728e-01 -8.35114397e-02
 -1.72715778e-02  6.84698838e-03  1.49895691e-01  1.15690266e-01
 -9.31451982e-02  1.42172430e-01 -2.04728865e-02  1.43920777e-01
  2.41279567e-03 -2.19571085e-01  5.56543720e-02  7.22320363e-01
 -6.83876344e-02 -5.22887195e-01 -1.25529772e-01 -7.32386288e-02
 -1.60454427e-02  5.05405950e-03  2.56319409e-01 -1.83231060e-01
 -1.53719274e-01 -7.70002337e-02 -2.69267206e-01 -2.82963453e-01
  1.59357966e-01  7.92660017e-02  7.94210793e-02  1.05406203e-01
 -3.08496963e-02 -4.41783363e-02  1.48101990e-01 -8.34899731e-02
 -2.98285345e-01  6.85567032e-02  2.65625358e-01 -5.07469924e-02
  5.08070082e-02 -2.67646154e-02 -3.51004916e-02  2.

### Case3.

In [300]:
lambda_vec = np.ones(103)
lambda_vec[6:] = 0.5
lambda_vec[-1] = 0
logist = mylogistic_l2(reg_vec = lambda_vec, max_iter = 100, tol = 1e-5, add_intercept = True)
logist.train2(x_train, y_train)
y_pred = logist.predict(x_test)

In [301]:
print("Accuracy:")
print(np.sum(y_pred[0] == y_test) / y_test.shape[0])
print("Weights:")
print(logist.weights)

Accuracy:
0.8476917967452674
Weights:
[ 3.34646216e-01  7.92280936e-02  2.39355963e-01  2.37284949e+00
  2.58821795e-01  3.53101158e-01  1.05205066e-01 -2.28732340e-02
  4.48252210e-02  5.15992831e-02 -1.07287833e-01 -4.29127032e-02
 -1.60925175e-01 -1.16218466e-01 -1.30551330e-01 -5.02840924e-02
 -5.49001987e-02 -8.04713506e-02 -1.46725490e-01 -1.04578683e-01
 -1.53582717e-03  1.71170836e-02  1.96478716e-01  1.41902524e-01
 -1.04370612e-01  1.78768280e-01 -4.41173454e-01  1.70410068e-01
  7.91220581e-03 -2.21157535e-01  5.58494437e-02  7.26012553e-01
 -6.88870877e-02 -5.24997320e-01 -1.26326590e-01 -7.39572924e-02
 -1.59789012e-02  5.11113561e-03  2.56408357e-01 -1.83269508e-01
 -1.53719118e-01 -7.69507197e-02 -2.69324041e-01 -2.84899112e-01
  1.59437687e-01  7.93046918e-02  7.94910901e-02  1.05464062e-01
 -3.08121338e-02 -4.74306038e-02  1.50482034e-01 -8.37453576e-02
 -2.97341376e-01  7.03067749e-02  2.64347542e-01 -5.07610656e-02
  5.08368187e-02 -2.67576443e-02 -3.51068702e-02  2.

### Q1.4
Further split the training data into subtraining (90%) and tuning (10%) to search for the best hyperparameters. Set the regularization coefficient for the constant term to zero. Allow different regularizations for continuous-valued and binary-valued features. Let  𝑎1  and  𝑎2  denote the regularization coefficients for continuous-valued and binary-valued features. Search the best  𝑎1  and  𝑎2  and report the test accuracy using the best hyper-parameters. You should follow the following procedure to search for the best hyperparameters.

Choose a set of grids among a reasonable range. For example, 10 grids in [0.01, 100].
Conduct grid search with the constraint that  𝑎1=𝑎2 . Record the best value  𝑎∗1  and  𝑎∗2 .  
Fix  𝑎1=𝑎∗1 , and search  𝑎2  for the best value, call the result the new  𝑎∗2 .  
Fix  𝑎2=𝑎∗2 , and search  𝑎1  for the best value.  
Report the selected  𝑎1  and  𝑎2 .  
Train a model using the selected hyper-parameters, and report the test accuracy.  

In [302]:
from sklearn.model_selection import train_test_split

#### 1. 
#### 2.

In [330]:
x_subtrain, x_tune, y_subtrain, y_tune = train_test_split(x_train, y_train, test_size=0.1)

In [320]:
lambdas = np.array(10 ** np.arange(-2, 2.4, 0.4))
acc = []

for l in lambdas:
    lambda_vec = np.ones(103) * l
    lambda_vec[-1] = 0
    logist = mylogistic_l2(reg_vec = lambda_vec, max_iter = 100, tol = 1e-5, add_intercept = True)
    logist.train2(x_subtrain, y_subtrain)
    y_pred = logist.predict(x_tune)
    acc.append(np.sum(y_pred[0] == y_tune) / y_tune.shape[0])

In [321]:
for i in range(len(lambdas)):   
    print("lambda:")
    print("{0:8.2f}".format(lambdas[i]))
    print("Accuracy:")
    print(acc[i])

lambda:
    0.01
Accuracy:
0.8434940855323021
lambda:
    0.03
Accuracy:
0.8434940855323021
lambda:
    0.06
Accuracy:
0.8434940855323021
lambda:
    0.16
Accuracy:
0.8434940855323021
lambda:
    0.40
Accuracy:
0.8434940855323021
lambda:
    1.00
Accuracy:
0.8434940855323021
lambda:
    2.51
Accuracy:
0.8434940855323021
lambda:
    6.31
Accuracy:
0.8434940855323021
lambda:
   15.85
Accuracy:
0.8430391264786169
lambda:
   39.81
Accuracy:
0.8421292083712466
lambda:
  100.00
Accuracy:
0.8439490445859873


Best $a_1$ is 100.  
#### 3.

In [322]:
lambdas = np.array(10 ** np.arange(-2, 2.4, 0.4))
acc = []

for l in lambdas:
    lambda_vec = np.ones(103) * 100
    lambda_vec[6:] = l
    lambda_vec[-1] = 0
    logist = mylogistic_l2(reg_vec = lambda_vec, max_iter = 100, tol = 1e-5, add_intercept = True)
    logist.train2(x_subtrain, y_subtrain)
    y_pred = logist.predict(x_tune)
    acc.append(np.sum(y_pred[0] == y_tune) / y_tune.shape[0])

In [323]:
for i in range(len(lambdas)):   
    print("lambda:")
    print("{0:8.2f}".format(lambdas[i]))
    print("Accuracy:")
    print(acc[i])

lambda:
    0.01
Accuracy:
0.8448589626933576
lambda:
    0.03
Accuracy:
0.8448589626933576
lambda:
    0.06
Accuracy:
0.8448589626933576
lambda:
    0.16
Accuracy:
0.8448589626933576
lambda:
    0.40
Accuracy:
0.8448589626933576
lambda:
    1.00
Accuracy:
0.8448589626933576
lambda:
    2.51
Accuracy:
0.8448589626933576
lambda:
    6.31
Accuracy:
0.8448589626933576
lambda:
   15.85
Accuracy:
0.8439490445859873
lambda:
   39.81
Accuracy:
0.8439490445859873
lambda:
  100.00
Accuracy:
0.8439490445859873


Best $a_2^*$ = 0.01
#### 4.

In [324]:
lambdas = np.array(10 ** np.arange(-2, 2.4, 0.4))
acc = []

for l in lambdas:
    lambda_vec = np.ones(103) * l
    lambda_vec[6:] = 0.01
    lambda_vec[-1] = 0
    logist = mylogistic_l2(reg_vec = lambda_vec, max_iter = 100, tol = 1e-5, add_intercept = True)
    logist.train2(x_subtrain, y_subtrain)
    y_pred = logist.predict(x_tune)
    acc.append(np.sum(y_pred[0] == y_tune) / y_tune.shape[0])

In [325]:
for i in range(len(lambdas)):   
    print("lambda:")
    print("{0:8.2f}".format(lambdas[i]))
    print("Accuracy:")
    print(acc[i])

lambda:
    0.01
Accuracy:
0.8434940855323021
lambda:
    0.03
Accuracy:
0.8434940855323021
lambda:
    0.06
Accuracy:
0.8434940855323021
lambda:
    0.16
Accuracy:
0.8434940855323021
lambda:
    0.40
Accuracy:
0.8434940855323021
lambda:
    1.00
Accuracy:
0.8434940855323021
lambda:
    2.51
Accuracy:
0.8434940855323021
lambda:
    6.31
Accuracy:
0.8439490445859873
lambda:
   15.85
Accuracy:
0.8430391264786169
lambda:
   39.81
Accuracy:
0.8425841674249318
lambda:
  100.00
Accuracy:
0.8448589626933576


Best $a_1^*$ is 100.  
#### 5.
$a_1^* = 100$  
$a_2^* = 0.01$  
#### 6.  

In [328]:
lambda_vec = np.ones(103) * 100
lambda_vec[6:] = 0.01
lambda_vec[-1] = 0
logist = mylogistic_l2(reg_vec = lambda_vec, max_iter = 100, tol = 1e-5, add_intercept = True)
logist.train2(x_train, y_train)
y_pred = logist.predict(x_test)

In [329]:
print("Accuracy:")
print(np.sum(y_pred[0] == y_test) / y_test.shape[0])
print("Weights:")
print(logist.weights)

Accuracy:
0.8467618731318499
Weights:
[ 3.19581105e-01  7.59955592e-02  6.65599609e-05  1.59511596e+00
  2.40627060e-01  3.33659584e-01  1.02148636e-01 -2.28821170e-02
  4.19673072e-02  5.20907393e-02 -1.04492231e-01 -4.52829050e-02
 -9.58160954e-02 -1.82260959e-01 -1.80675502e-01 -7.29394798e-02
 -1.10210956e-01 -1.44627598e-01 -2.20755461e-01 -1.66491253e-01
  2.69809334e-02  3.22435181e-02  2.88337758e-01  1.99105361e-01
 -1.58517029e-01  2.57095845e-01 -2.72772071e-01  2.25696122e-01
 -3.84933558e-03 -2.16959892e-01  5.41119174e-02  7.12739117e-01
 -6.69043661e-02 -5.15630735e-01 -1.25035331e-01 -7.15583271e-02
 -1.71524835e-02  3.38165186e-03  2.51357757e-01 -1.76985633e-01
 -1.54710769e-01 -7.79658560e-02 -2.73355152e-01 -1.94297654e-01
  1.56996865e-01  7.57566833e-02  7.70939585e-02  1.01663888e-01
 -3.16554916e-02 -4.74290865e-02  1.52304670e-01 -8.62562847e-02
 -2.89586837e-01  6.54654637e-02  2.56602940e-01 -4.59148004e-02
  5.01816686e-02 -2.44539562e-02 -3.59018960e-02  1.

### Q1.5
Use sklearn.linear_model.LogisticRegression to train and test the model (including hyperparameter tuning). Compare the estimated parameters and test accuracy with those from your own models.

In [332]:
import sklearn.linear_model as sk

In [335]:
lambdas = np.array(10 ** np.arange(-2, 2.4, 0.4))
acc = []

for l in lambdas:
    clf = sk.LogisticRegression(C = l, solver = "newton-cg")
    clf.fit(x_subtrain, y_subtrain)
    acc.append(clf.score(x_tune, y_tune))

In [336]:
for i in range(len(lambdas)):   
    print("lambda:")
    print("{0:8.2f}".format(lambdas[i]))
    print("Accuracy:")
    print(acc[i])

lambda:
    0.01
Accuracy:
0.8448275862068966
lambda:
    0.03
Accuracy:
0.8444960212201591
lambda:
    0.06
Accuracy:
0.8454907161803713
lambda:
    0.16
Accuracy:
0.8454907161803713
lambda:
    0.40
Accuracy:
0.8458222811671088
lambda:
    1.00
Accuracy:
0.8461538461538461
lambda:
    2.51
Accuracy:
0.8461538461538461
lambda:
    6.31
Accuracy:
0.8461538461538461
lambda:
   15.85
Accuracy:
0.8461538461538461
lambda:
   39.81
Accuracy:
0.8461538461538461
lambda:
  100.00
Accuracy:
0.8461538461538461


Best C is 100.

In [338]:
clf = sk.LogisticRegression(C = 100, solver = "newton-cg")
clf.fit(x_train, y_train)
print(clf.score(x_test, y_test))
print(clf.coef_)

0.8476917967452674
[[ 3.34833249e-01  7.92629434e-02  3.79432917e-01  2.38877195e+00
   2.59118209e-01  3.53328115e-01  1.05926042e-01 -2.19609210e-02
   4.65158719e-02  5.22888068e-02 -1.06352706e-01 -4.21661704e-02
  -2.36770572e-01 -7.73035706e-02 -9.69778585e-02 -3.59500929e-02
  -2.25763356e-02 -4.12737767e-02 -9.98654047e-02 -6.87275977e-02
  -1.77459781e-02  9.91423091e-03  1.42444748e-01  1.07662845e-01
  -6.94402638e-02  1.33660935e-01 -6.38406921e-01  1.36675699e-01
   1.61121575e-02 -2.22877062e-01  5.60868563e-02  7.30157855e-01
  -6.94353539e-02 -5.27439844e-01 -1.27194498e-01 -7.47245806e-02
  -1.59077239e-02  5.19774824e-03  2.56629270e-01 -1.83320559e-01
  -1.53743102e-01 -7.69370283e-02 -2.69331102e-01 -2.88605809e-01
   1.59563788e-01  7.94219265e-02  7.96273315e-02  1.05583093e-01
  -3.07429216e-02 -5.08809857e-02  1.52918302e-01 -8.38144982e-02
  -2.96420486e-01  7.21717207e-02  2.63137641e-01 -5.08722623e-02
   5.08624758e-02 -2.67858981e-02 -3.51245310e-02  2.0586

Accuracy given by sklearn's logistic regression model is 0.8476917967452674.  
Mine is 0.8476917967452674.  
Both are almost identical.